### DISCUSSION TITLE
Ideas for improvements in JustPy welcome!

### DISCUSSION THREAD

**Discussion by elimintz** [elimintz/justpy#189] [Ideas]
Please feel free to discuss any idea you have, even if it is not completely thought out.

**Comment by elimintz** [comment]
JavaScript is very much integrated with HTML. Could a variant of JustPy be an HTML format file in which Python replaces JavaScript? Would that be more intuitive for some users?

**Reply by mcweekly** [reply]
Is this something like the Hotwire approach? https://hotwire.dev/

**Reply by elimintz** [reply]
I was thinking more of the alpine.js approach https://github.com/alpinejs/alpine

**Reply by 99hats** [reply]
Ooh, that's really nice. The declarative nature of it reminds me of intercooler.js which apparently just went 2.0 - https://htmx.org/

**Reply by elimintz** [reply]
Thanks, didn't know about htmx. It is also a very interesting concept.

**Reply by upascal** [reply]
Yes! I agree. Would something like this be possible with justpy's html parse feature?

**Reply by platinops** [reply]
This sounds a bit like what Microsoft is doing with [Blazor](https://en.m.wikipedia.org/wiki/Blazor), letting developers write HTML+C# only.

```html
<h1>Blazor code example</h1>
<p>count: @count</p>
<button class="btn btn-primary" @onclick="IncCount">Click to increment</button>

@code {
  private int count = 0;

  private void IncCount()
  {
    count++;
  }
}
```
Due to my lack of JS knowledge, I personally would prefer a pure Python approach like Justpy and Dash, where minimal to no HTML and JS is required. (I prefer Justpy a lot over Dash, though the latter does seem to be more actively maintained.)

What I do like about Blazor is that it only sends a diff of the webpage from server to client, which makes it more efficient than Justpy and Vue where I understood the whole page is being sent up and down on each event that triggers Python code in the server end. If a similar approach is possible on Justpy, that would be awesome!

**Comment by 99hats** [comment]
A plugin system might inspire an entire ecosystem.

**Comment by ChronoDK** [comment]
It would be neat if it was build for FastAPI. The Webpage class and HTML components stuff is so cool, but all the routing stuff etc. seems like a big undertaking that FastAPI already handles well.

**Reply by elimintz** [reply]
JustPy supports path parameters: https://justpy.io/tutorial/request_object/#path-parameters
What additional features would you like to see?

**Comment by bigfatwhale** [comment]
I have been looking at the electron framework and wonder if justpy could be made to support the generation of code for electron framework as well. For example, if we have a file dialog that can access the local file system, then justpy would need to generate code for the ipcMain side of things?

**Reply by elimintz** [reply]
This is an improvement that could happen in the future when using the Quasar components. It is possible to build Quasar applications for electron, but this is not yet supported by JustPy

**Comment by grigolet** [comment]
It would be really nice if justpy could handle POST requests (please correct me if it does because I just found out this amazing framework today). I was planning on trying it for creating a detector control system but it would need handling requests from entities that are not users/browser

**Reply by elimintz** [reply]
Hi, 
JustPy is useful for creating user interfaces and does not support POST requests. Perhaps for what you need, https://fastapi.tiangolo.com/ would be better. That is what I use when I need an API. Both fastapi and JustPy are based on starlette and can work together.

**Reply by grigolet** [reply]
Thanks, I see. Perhaps I can give you a particular use case so you could tell me if you think FastAPI would be nevertheless a better solution. I have an external service (say like grafana, kibana or similar) that is sending a POST request whenever an alert is received. I would like that everytime this POST request is sent a UI of the page the user is browsing gets modified. Regarding FastAPI and justpy is there any code sample around showing how they can work together?

**Reply by elimintz** [reply]
Is the message specific to each user?

**Reply by grigolet** [reply]
It should be general, not depending on the user

**Comment by mwx23** [comment]
It would be cool to have a concept of a store like React Redux. I find myself creating nested components and having to pass down the model. I also may be doing that completely wrong, I've never used Vue before. But the idea of having a global store that components get their state from, and also some kind of event bus where a component could issue an action/event that get actioned by the store or perhaps a component that is listening for that action/event.

**Reply by elimintz** [reply]
What I do, and I am not sure it is applicable for all applications, is just use a global python dict as a global store. Since there is no backend/frontend distinction in JustPy it works quite well. 

I haven't yet needed to use an event bus. It would help if you have some particular example in mind where that would be useful.

**Comment by matthoendorf** [comment]
I realize the time committment that this project requires. Have you considered enabling sponsorships? Looking forward to the next release!

**Comment by elimintz** [comment]
Unfortunately, the sponsorships will not create more time. The problem is the testing and making sure I don't break anything. But I am getting there.

**Reply by matthoendorf** [reply]
Completely understand. Even so, I'd love to sponsor as a thanks for your time.

**Reply by elimintz** [reply]
Very much appreciated! There is really no need. The sentiment is enough.
